ytypes/node.go

```diff
@@ -399,6 +401,9 @@ type GetOrCreateNodeOpt interface {
 // along the path if they are nil.
 // Function returns the value and schema of the node as well as error.
 // Note that this function may modify the supplied root even if the function fails.
+// Note that this function may create containers or list entries even if the input path is a shadow path.
+// TODO(wenbli): a traversal should remember what containers or list entries were created so that a shadow path
+//               can later undo this. This applies to SetNode as well.
```

Comment:
If a container == nil, or a map has no members, is there a reason to delete it? I presume that any API (`GetNode`, `Diff`, `TogNMINotifications`, `Marshal7951`) all have to handle this case anyway?

It's the same as doing `&oc.Device{}.GetOrCreateSystem().GetOrCreateNtp()` -- in this scenario even though there was nothing to be marshalled, we set the containers to non-nil.

Reply:
I thought about whether this matters and my concern is that if you do `&oc.Device{}.GetOrCreateInterface("eth0")`, the key leaf gets populated. Otherwise I agree there is no reason to delete it.

Reply:
I meant if you pass in the path `/interfaces/interface[name="eth0"]` to `GetOrCreateNode`.

Reply:
The key leaf gets created, and its value set though right, so at this point the `map` entry is not actually empty. 

 I think the context that I was missing here is why we need this for shadow paths but not for the normal paths?

Reply:
There is a short explanation of the purpose of shadow paths where the flag is defined: https://github.com/openconfig/ygot/blob/master/generator/generator.go#L72

Reply:
Sorry -- I understand what a shadow path is. Why do we delete empty entries for shadow paths and not for normal paths though? i.e., what is the problem that you're trying to solve here, and why is it only a problem for shadow paths?

Reply:
The problem is that even if the function call didn't fail for a shadow path, it shouldn't have created a list entry since it made it not empty. Unmarshalling a shadow path should not modify the struct. See above for an explanation, happy to discuss this offline if it's still unclear.

Reply:
Per my other response, I'm not sure that we're aligned on when these shadow paths can occur, and what needs to be handled -- let's get to the bottom of that and then come back to this thread.